### ISSUE TITLE
Thumbnail oversampling

### ISSUE THREAD

**Issue by timhoffm** [sphinx-gallery/sphinx-gallery#717]
https://github.com/sphinx-gallery/sphinx-gallery/blob/353ff9f972c240e3855e91283a89dddef669adf9/sphinx_gallery/gen_gallery.py#L64

What is the reason for 2.5x oversampling? We had a recent PR for reducing that for Matplotlib specifically (https://github.com/matplotlib/matplotlib/pull/17275). We were assuming that 2x is enough, and one may even get sharper pictures on non-scaled screens.

**Comment by larsoner** [comment]
Digging a ways back into the history, I'm guessing it's because `sklearn` originally used the thumbnails as part of some carousel:

https://github.com/scikit-learn/scikit-learn/blob/0.16.X/doc/sphinxext/gen_rst.py#L941-L945

I agree this is probably not good for most projects nowadays, in my projects I tend to set `'thumbnail_size': (160, 112)` or similar. Would a big doc change recommending that this be set differently be enough here? Otherwise we probably need to do a deprecation cycle for people who haven't explicitly set it, which might be okay.

cc @GaelVaroquaux @lesteve in case you know more details...

**Comment by timhoffm** [comment]
Thanks for the explanation. For now, I‘ll set the size manually. From a user‘s perspective, the default should be reasonable. There is little point in recommending everybody to change a standard setting. IMHO this is worth the deprecation effort.

**Comment by larsoner** [comment]
@lucyleeow want to take a stab at this? Basically if `thumbnail_size` is not given by the user then it should `warnings.warn(..., DeprecationWarning)` and for now use the old value, and in the next version we'll remove the warning and use the new value

**Comment by lucyleeow** [comment]
Sorry just to clarify, if `thumbnail_size` is the default value (`(400, 280)`), give the warning?

What would be the new value for the next version? '160, 112' ? 

I'm also confused by our documentation:
> The gallery uses various CSS classes to display these thumbnails, which default to maximum 160x160px. 

My CSS isn't good but I don't know where in `gallery.css` this is set?
```
.sphx-glr-thumbcontainer img {
  display: inline;
  max-height: 112px;
  max-width: 160px;
}
```
My understanding is that the `thumbnail_size` will be scaled to within max-height and max-width, while maintaining aspect ratio specified in `thumbnail_size`.

**Comment by larsoner** [comment]
> Sorry just to clarify, if thumbnail_size is the default value ((400, 280)), give the warning?

No, only emit the DeprecationWarning if they have not set `thumbnail_size` in `conf.py`. People with the value not explicitly set are the only ones who will be affected by the change in default value

> My understanding is that the thumbnail_size will be scaled to within max-height and max-width, while maintaining aspect ratio specified in thumbnail_size.

Check the code -- I think we might actually letterbox/pillarbox the image to make it actually be `thumbnail_size`. And from a recent MNE build where we set `thumbnail_size` both the missing-img and actual images are pillarboxed:

![sphx_glr_plot_10_epochs_overview_thumb](https://user-images.githubusercontent.com/2365790/85434777-8dda5d80-b554-11ea-9cc0-01ff10a6d33a.png)

![sphx_glr_plot_stats_cluster_spatio_temporal_thumb](https://user-images.githubusercontent.com/2365790/85434803-992d8900-b554-11ea-931e-814983ef3b0b.png)

**Comment by lucyleeow** [comment]
> No, only emit the DeprecationWarning if they have not set `thumbnail_size` in `conf.py`

Sorry, but how do I check if it's been set in `conf.py`? 

Okay I see we do scale the image to `thumbnail_size`:
https://github.com/sphinx-gallery/sphinx-gallery/blob/aae7049e4eac4b52f1a53bca23b5f63afdb20d56/sphinx_gallery/gen_rst.py#L308

Just to confirm, the letterbox/pillarbox happens when the `thumbnail_size` does not have the same aspect ratio as (400, 280) ?

Also what does the thumbnail CSS in `gallery.css` do then...?

(thank you)

**Comment by larsoner** [comment]
> Sorry, but how do I check if it's been set in conf.py?

See for example

https://github.com/sphinx-gallery/sphinx-gallery/blob/master/sphinx_gallery/gen_gallery.py#L108-L115

> Just to confirm, the letterbox/pillarbox happens when the thumbnail_size does not have the same aspect ratio as (400, 280) ?

Yes indeed

**Comment by lucyleeow** [comment]
Sorry to bother, last question. From the `gallery.css` i interpret this:
```
.sphx-glr-thumbcontainer img {
  display: inline;
  max-height: 112px;
  max-width: 160px;
}
```
as setting maximum 160x112px but the doc says we set maximum 160x160px?

**Comment by larsoner** [comment]
Docs are probably wrong on that point

**Comment by timhoffm** [comment]
1. Have you considered 320x224 (i.e. 2x CSS values) as the default? This will cover Retina/HiDPI displays and easily scales down for regular displays without much artifacts. The downside is the file size. It‘s a trade-off, but I would go for the better looking option as the default.

2. I propose not to add a deprecation warning to the code. I assume that most users don‘t care explicitly about the thumbnail size. They just want sphinx-gallery to do „the right thing“ without them having to configure it explicitly. When adding the deprecation warning, you force everybody to set the thumbnail size explicitly to silence that warning. Apart from burdening the users with work, this locks them in on a manual setting that will later become default and thus obsolete. This is problematic as any future default changes don‘t take effect for them (maybe you decide to go with 1x CSS for now but want to change that to 2x CSS later, or you want to change the default layout for larger previews in the future).
I think there are very few people explicitly wanting the current default. Given the hassle a deprecation warning makes for all other users, it’s better to only document the planned API change in the release notes 1-2 releases before it takes effect.

**Comment by lucyleeow** [comment]
Good points.

I agree with 2 as well since the change won't really affect people (unless they have a really resolution screen and really look at the thumbnails) and people interested in changing can read the docs. I will amend my PR and see what @larsoner says.

**Comment by timhoffm** [comment]
Closed by #719.
